### ISSUE TITLE
API guidelines for asynchronous libraries

### ISSUE THREAD

**Issue by jonhoo** [tokio-rs/doc-push#79]
The [Rust API guidelines](https://rust-lang-nursery.github.io/api-guidelines/) are a initiative for making Rust libraries feel more similar. Something similar should exist for asynchronous libraries. Should my methods take `&self` or `self`? Should I return `(Self, X)` or `(X, Self)`? How do I return `Self` on error if I take `&self`?

**Comment by carllerche** [comment]
I don’t know if we want to be as strict as the rust api guidelines or is we would have as much content . Maybe this would work as a page in the cookbooks section?

**Comment by jonhoo** [comment]
I don't think it needs to be quite as strict, but some guidelines/best practices I think would help a lot. Might also encourage some more interface uniformity in the ecosystem.
